opensearch_single_kernel/managers/lock.py

```diff
@@ -0,0 +1,507 @@
+#!/usr/bin/env python3
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""OpenSearch Lock manager.
+
+Ensure that only one node (re)starts, joins the cluster, or leaves the cluster at a time.
+
+The workflow logic goes alongside the following:
+
+1. If there are opensearch online nodes:
+   a) the node requesting the lock attempts to create a doc with the name of the node
+   b) if it succeeds => the unit gets the lock
+   c) if it fails => the unit doesn't get the lock as it is held by another unit
+   d) when the unit completes with their locked operation => releases the lock => deletes the doc
+2. if there are no online nodes:
+   a) we make use of a flag in the relation data
+   b) we check on the existence of the flag to know if the lock is held or not
+   c) if not there => we set the lock (flag in the peer rel data)
+   d) we release the lock by removing that flag from the rel data
+"""
+import json
+import os
+from typing import List, Optional
+
+import ops
+
+from opensearch_single_kernel.common.client import OpenSearchClient
+from opensearch_single_kernel.common.constants import DeploymentType, Scope, StartMode
+from opensearch_single_kernel.common.exceptions import OpenSearchHttpError
+from opensearch_single_kernel.core.models import DeploymentDescription
+from opensearch_single_kernel.core.state import ClusterState
+from opensearch_single_kernel.managers.base import BaseManager
+from opensearch_single_kernel.utils.helpers import format_unit_name
+from opensearch_single_kernel.utils.topology import ClusterTopology
+from opensearch_single_kernel.workload.base import BaseWorkload
+
+
+class PeerLockManager(BaseManager):
+    """Fallback lock when all units of OpenSearch are offline."""
+
+    def __init__(self, state: ClusterState, workload: BaseWorkload):
+        self.state = state
+        self.workload = workload
```

Comment:
```suggestion
      super().__init__(state, workload)
```
We can benefit from super.